File: Application/Sources/Helpers/Categories/SRGMedia+PlaySRG.m
//
//  Copyright (c) SRG SSR. All rights reserved.
//
//  License information is available from the LICENSE file.
//

#import "SRGMedia+PlaySRG.h"

@import libextobjc;

@implementation SRGMedia (PlaySRG)

- (BOOL)play_isToday
{
    return [NSCalendar.srg_defaultCalendar isDateInToday:self.date];
}

- (NSString *)play_fullSummary
{
    if (self.lead.length && self.summary.length && ![self.summary containsString:self.lead]) {
        return [NSString stringWithFormat:@"%@\n\n%@", self.lead, self.summary];
    }
    else if (self.summary.length) {
        return self.summary;
    }
    else if (self.lead.length) {
        return self.lead;
    }
    else {
        return nil;
    }
}

- (BOOL)play_areSubtitlesAvailable
{
    return self.play_subtitleVariants.count != 0;
}

- (BOOL)play_isAudioDescriptionAvailable
{
    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"%K == %@", @keypath(SRGVariant.new, type), @(SRGVariantTypeAudioDescription)];
    return [self.play_audioVariants filteredArrayUsingPredicate:predicate].count != 0;
}

- (BOOL)play_isMultiAudioAvailable
{
    NSArray<NSLocale *> *locales = [self.play_audioVariants valueForKey:@keypath(SRGVariant.new, locale)];
    return [NSSet setWithArray:locales].count > 1;
}

- (BOOL)play_isWebFirst
{
    NSDate *date = NSDate.date;
    return [self.date compare:date] == NSOrderedDescending && [self timeAvailabilityAtDate:date] == SRGTimeAvailabilityAvailable && self.contentType == SRGContentTypeEpisode;
}

- (NSArray<NSString *> *)play_subtitleLanguages
{
    return [self.play_subtitleVariants valueForKeyPath:@keypath(SRGVariant.new, language)];
}

- (NSArray<NSString *> *)play_audioLanguages
{
    return [self.play_audioVariants valueForKeyPath:@keypath(SRGVariant.new, language)];
}

- (NSArray<SRGVariant *> *)play_subtitleVariants
{
    return [self subtitleVariantsForSource:self.recommendedSubtitleVariantSource];
}

- (NSArray<SRGVariant *> *)play_audioVariants
{
    return [self audioVariantsForSource:self.recommendedAudioVariantSource];
}

@end

#pragma mark Functions

BOOL PlayIsSwissTXTURN(NSString *mediaURN)
{
    return [mediaURN containsString:@":swisstxt:"];
}

File: Application/Sources/Helpers/Extensions/SRGMedia+PlaySRG.swift
//
//  Copyright (c) SRG SSR. All rights reserved.
//
//  License information is available from the LICENSE file.
//

import SRGDataProviderModel

extension SRGMedia {
    // FIXME: Ask IL to get this information in a media
    /**
     *  Return `true` iff the URN is related to a live center event.
     */
    @objc static func PlayIsSwissTXTURN(_ mediaURN: String) -> Bool {
        mediaURN.contains(":swisstxt:")
    }

    var play_isToday: Bool {
        NSCalendar.srg_default.isDateInToday(date)
    }

    // Return a concatenation of lead and summary, iff summary not contains the lead, to avoid duplicate information.
    @objc var play_fullSummary: String? {
        if let lead, !lead.isEmpty, let summary, !summary.isEmpty, !summary.contains(lead) {
            "\(lead)\n\n\(summary)"
        } else if let summary, !summary.isEmpty {
            summary
        } else if let lead, !lead.isEmpty {
            lead
        } else {
            nil
        }
    }

    var play_summary: String? {
        leadOrSummary
    }

    private var leadOrSummary: String? {
        lead?.isEmpty ?? true ? summary : lead
    }

    var play_areSubtitlesAvailable: Bool {
        !play_subtitleVariants.isEmpty
    }

    var play_isAudioDescriptionAvailable: Bool {
        play_audioVariants.contains(where: { $0.type == .audioDescription })
    }

    var play_isMultiAudioAvailable: Bool {
        let locales = play_audioVariants.map(\.locale)
        return Set(locales).count > 1
    }

    @objc var play_isWebFirst: Bool {
        date > Date() && timeAvailability(at: Date()) == .available && contentType == .episode
    }

    var play_subtitleLanguages: [String] {
        play_subtitleVariants.map { $0.language ?? $0.locale.identifier }
    }

    var play_audioLanguages: [String] {
        play_audioVariants.map { $0.language ?? $0.locale.identifier }
    }

    private var play_subtitleVariants: [SRGVariant] {
        subtitleVariants(for: recommendedSubtitleVariantSource) ?? []
    }

    private var play_audioVariants: [SRGVariant] {
        audioVariants(for: recommendedAudioVariantSource) ?? []
    }

    var publicationDate: Date {
        startDate ?? date
    }
}
